concurrency/DCMetricsUtils.java

```diff
@@ -0,0 +1,481 @@
+            METRIC_INDEX.remove(metricId);
+            TIMER_SAMPLE_INDEX.remove(metricId);
+        }
+
+        return this;
+    }
+
+    /**
+     * Removes a metric from the current map
+     * @param metricId The ID of the metric that you want to remove
+     * @return This instantiation of DCMetricsUtil
+     */
+    public DCMetricsUtils removeMetric(String metricId) {
+        Slog.v(TAG, "Remove Metric: " + metricId);
+
+        METRIC_INDEX.remove(metricId);
+        TIMER_SAMPLE_INDEX.remove(metricId);
+
+        return this;
+    }
+
+    /**
+     * Detects whether or not a metric currently exists in the metric table
+     * @param metricId The ID of the metric that you want to check on
+     * @return Whether or not it exists in the current table
+     */
+    public boolean containsMetric(String metricId) {
+        Slog.v(TAG, "Validate Containment of Metric: " + metricId);
+
+        return METRIC_INDEX.containsKey(metricId);
+    }
+
+    /**
+     * In a given MetricEvent in the METRIC_INDEX, increment a counter by the given amount.
+     * If the MetricEvent doesn't exist, it will be created and added it to the index first.
+     */
+    public DCMetricsUtils incrementMetricCounter(String metricId, String counterName, double increment) {
+
+        Slog.v(TAG, "Increment Metric Counter for Metric: " + metricId);
+
+        if (!containsMetric(metricId)) {
+            createMetric(metricId);
+        }
+
+        MetricEvent metricEvent = getMetricEvent(metricId);
+        if (metricEvent != null) {
+            counterName = stripIllegalCharacters(counterName);
+            metricEvent.addCounter(counterName, increment);
+        }
+        return this;
+    }
+
+    /**
+     * Adds a timer for a specific metric id
+     * @param metricId The ID to the metric that needs the timer
+     * @param timerName The name of the metric timer
+     * @param startTime The time that you want to start the timer at, or add on to the current timer
+     * @param recurrences The amount of samples taken to sum up the given startTime
+     * @return This instantiation of DCMetricsUtil
+     */
+    public DCMetricsUtils createMetricTimer(String metricId, String timerName, long startTime, int recurrences) {
+        MetricEvent metricEvent = METRIC_INDEX.get(metricId);
+
+        if (metricEvent == null) {
+            Slog.w(TAG, "Adding timer for a metric that doesn't exist. " + metricId);
+            return this;
+        }
+
+        timerName = stripIllegalCharacters(timerName);
+        metricEvent.addTimer(timerName, startTime, recurrences);
+
+        Map<String, Integer> timerMap = TIMER_SAMPLE_INDEX.get(metricId);
+        if (timerMap != null) {
+            timerMap.put(timerName, 0);
+        } else {
+            ConcurrentHashMap<String, Integer> newTimerMap = new ConcurrentHashMap<>();
+            newTimerMap.put(timerName, 0);
+            TIMER_SAMPLE_INDEX.put(metricId, newTimerMap);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **225, 226, and 231**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).